server/game/core/ability/abilityTargets/TargetResolver.ts

```diff
@@ -0,0 +1,72 @@
+import { ITargetResolverBase } from '../../../TargetInterfaces';
+import { AbilityContext } from '../AbilityContext';
+import * as Contract from '../../utils/Contract';
+import { GameSystem } from '../../gameSystem/GameSystem';
+import PlayerOrCardAbility from '../PlayerOrCardAbility';
+import { RelativePlayer, Stage } from '../../Constants';
+
+/**
+ * Base class for all target resolvers.
+ */
+export abstract class TargetResolver<TProps extends ITargetResolverBase<AbilityContext>> {
+    protected dependentTarget = null;
+    protected dependentCost = null;
+
+    public constructor(protected name: string, protected properties: TProps, ability: PlayerOrCardAbility) {
+        if (this.properties.dependsOn) {
+            const dependsOnTarget = ability.targetResolvers.find((target) => target.name === this.properties.dependsOn);
+
+            // assert that the target we depend on actually exists
+            Contract.assertNotNullLike(dependsOnTarget);
+            // TODO: Change the dependsOn system to allow multiple dependent targets.
+            if (dependsOnTarget.dependentTarget != null) {
+                throw new Error('Attempting to set a dependent target for source target <insert target name here> but it already has one. Having multiple dependent targets for the same source target has not yet been implemented');
+            }
+
+            dependsOnTarget.dependentTarget = this;
+        }
+    }
+
+    protected abstract hasLegalTarget(context): boolean;
+
+    protected abstract getAllLegalTargets(context: AbilityContext): any[];
+
+    protected abstract checkTarget(context: AbilityContext): boolean;
+
+    protected abstract hasTargetsChosenByInitiatingPlayer(context: AbilityContext): boolean;
+
+    protected canResolve(context) {
+        // if this depends on another target, that will check hasLegalTarget already
+        return !!this.properties.dependsOn || this.hasLegalTarget(context);
+    }
+
+    protected resolve(context: AbilityContext, targetResults, passPrompt = null) {
```

Comment:
This is a good start! One more big suggestion I have.

Typically when you provide a method like this in a base class, it's doing one of two things:

1. It provides some basic default functionality which is enough for at least some of the derived classes to use without modifying / overriding
2. It makes calls into abstract methods that the derived classes are required to provide implementations for, thus each derived class can specialize its functionality to their own case

This pattern you have here of requiring derived classes to overload but also call super() is a little tricky because it's not super clear that that's the intent and it's also not enforceable by the transpiler the same way that `abstract` is.

An easy way to fix this via (2) above is to, instead of overriding and calling `super.resolve()`, add a new abstract method named `resolveInternal` or `resolveImpl` or something and call that at the end of this method, and then the child classes override that.

I do think though you could go further with that pattern, there is still a lot of code shared between the `resolve()` methods of the derived classes that could be pulled in here. The 'pay costs first' button, the waiting prompt, etc are all  in common.

It might require making a couple more abstract methods so that different pieces of the pipeline can be implemented by subclasses - e.g., you might need like `abstract buildChoices()`, `abstract buildHandlers()`, stuff like that. But at the end of the day it would make the whole set of them much easier to understand and maintain since we could modify all common behaviors in the base class instead of needing to duplicate changes across all of them.

Reply:
Thank you for the guidance. It feels obvious now but I don't think I can remember doing a refactor like this(making a base class and pulling out shared code) in recent memory so I didn't really think of making new functions.

Reply:
Of course! Yeah I don't know that we've done it much in this repo since most methods are pretty small and self-contained.